### PR TITLE
Add new guide how to create a plugin - Closes #888

### LINKED ISSUE #888: Add guide how to create a custom plugin
Parent issue: #841

### PR DESCRIPTION
**Pull request by Tschakki**
Closes #888

### CODE REVIEW

File: modules/ROOT/pages/guides/app-development/plugin.adoc [hunk 1]
```diff
@@ -0,0 +1,386 @@
...
+}
+
+module.exports = MyPlugin;
+----
+
+== Defining the plugin configuration
+
+A plugin can be configured by setting the related properties in the application xref:{url_architecture}[configuration].
+
+What properties should bbe added to the application config is defined in the `defaults()` getter.
```
Comment (mitsuaki-u) [inline modules/ROOT/pages/guides/app-development/plugin.adoc]:
```suggestion
What properties should be added to the application config is defined in the `defaults()` getter.
```

Comment (mitsuaki-u) [reply modules/ROOT/pages/guides/app-development/plugin.adoc]: The wording for this sentence seems off

File: modules/ROOT/pages/guides/app-development/plugin.adoc [hunk 2]
```diff
@@ -0,0 +1,386 @@
...
+
+For a better overview, create a new file `defaults.js` and define the schema for the configuration options here like in the example below:
```
Comment (mitsuaki-u) [inline modules/ROOT/pages/guides/app-development/plugin.adoc]:
```suggestion
For a better overview, create a new file `defaults.js` and define the schema for the configuration options here as in the example below:
```

File: modules/ROOT/pages/guides/app-development/plugin.adoc [hunk 3]
```diff
@@ -0,0 +1,386 @@
...
+The `unload()` method contains the logic that needs to be executed to unload the plugin correctly.
+
+The `channel`, which is available inside of the `load()` function, allows to access the xref:{url_rpc_endpoints}[] in order to subscribe to events or to invoke certain actions within the application, to retrieve the desired data.
+
+In this example, we subscribe to the event `app:transaction:new`, which is published everytime a new transaction is added to the application.
+Next, the transaction is decoded and checked for it's `moduleID` and `assetID`.
+If the transaction is a register delegate transaction, the delegate name is saved under `this._latestDelegate` and a new event `myPlugin:newDelegate` is published, which is announcing the new delegate to the application.
+
+Additionally, we subscribe to the event `app:block:new`, which is published everytime a new block is added to the blockchain.
+Next, the block is decoded and the timestamp of the block is pushed into the `_knownTimestamps` array.
```
Comment (mitsuaki-u) [inline modules/ROOT/pages/guides/app-development/plugin.adoc:217]: what is the purpose of `_knownTimestamps`?

File: modules/ROOT/pages/guides/app-development/plugin.adoc [hunk 4]
```diff
@@ -0,0 +1,386 @@
...
+
+Similar to modules, plugins expose `actions` and `events`, which are interfaces that allow other plugins or external services to interact with the plugin.
+
+In this example, two events are added:
+
+* `newDelegate`, which is published in the `load()` function, when a new delegate is registered in ther network.
+* `timestamp`, which is published in the `load()` function, when a new block is added to the blockchain.
+
+In addition, two actions are added:
+
```
Comment (mitsuaki-u) [inline modules/ROOT/pages/guides/app-development/plugin.adoc:292]: The purpose of this example plugin is a little bit confusing. It is storing the latest delegate and also storing the timestamps for each new block. Are they related somehow?

**Review by mitsuaki-u — CHANGES_REQUESTED** [review]

File: modules/ROOT/pages/guides/app-development/plugin.adoc (hunk 3 above)
Comment (Tschakki) [reply modules/ROOT/pages/guides/app-development/plugin.adoc:217]: it is a list of timestamps of the blocks that where added to the chain, while the plugin was active.

File: modules/ROOT/pages/guides/app-development/plugin.adoc (hunk 1 above)
Comment (Tschakki) [reply modules/ROOT/pages/guides/app-development/plugin.adoc]:
```suggestion
Which properties are added to the application config by the plugin is defined in the `defaults()` getter.
```

Better like this?

Comment (Tschakki) [reply modules/ROOT/pages/guides/app-development/plugin.adoc]:
```suggestion
The `defaults()` getter defines which properties are added to the application config by the plugin.
```

Or this.

File: modules/ROOT/pages/guides/app-development/plugin.adoc (hunk 4 above)
Comment (Tschakki) [reply modules/ROOT/pages/guides/app-development/plugin.adoc:292]: The purpose is to give an example how to create a plugin, and I wanted to include common examples. Decoding blocks and transactions are both pretty common and relevant examples, that's why I included them here both. Do you maybe have other suggestions?

**Review by Incede — DISMISSED** [review]

File: modules/ROOT/pages/guides/app-development/plugin.adoc [hunk 5]
```diff
@@ -0,0 +1,386 @@
...
+:url_tutorials_hello: tutorials/hello-world.adoc
+
+This guide explains step-by-step, how to create a custom plugin for a blockchain application built with the Lisk SDK.
+
+.Prerequisites
+[NOTE]
+====
+To follow this guide, the following is assumed:
+
+* a basic blockchain application is already initiated, like explained in the previous guide xref:{url_guides_setup}[].
```
Comment (chris529) [inline modules/ROOT/pages/guides/app-development/plugin.adoc]:
```suggestion
* a basic blockchain application is already initiated, as explained in the previous guide xref:{url_guides_setup}[].
```

File: modules/ROOT/pages/guides/app-development/plugin.adoc [hunk 6]
```diff
@@ -0,0 +1,386 @@
...
+:url_modules_dpos: dpos-module.adoc
+:url_rpc_endpoints: rpc-endpoints.adoc
+:url_tutorials_hello: tutorials/hello-world.adoc
+
+This guide explains step-by-step, how to create a custom plugin for a blockchain application built with the Lisk SDK.
+
+.Prerequisites
+[NOTE]
+====
+To follow this guide, the following is assumed:
```
Comment (chris529) [inline modules/ROOT/pages/guides/app-development/plugin.adoc]:
```suggestion
To use this guide, it is assumed that the following criteria have been met:
```

File: modules/ROOT/pages/guides/app-development/plugin.adoc [hunk 7]
```diff
@@ -0,0 +1,386 @@
...
+module.exports = MyPlugin;
+----
+
+== Defining the plugin configuration
+
+A plugin can be configured by setting the related properties in the application xref:{url_architecture}[configuration].
+
+The `defaults()` getter defines which properties are added to the application config by the plugin.
+
+For a better overview, create a new file `defaults.js` and define the schema for the configuration options here as in the example below:
```
Comment (chris529) [inline modules/ROOT/pages/guides/app-development/plugin.adoc]:
```suggestion
For a better overview, create a new file `defaults.js` and define the schema for the configuration options as shown in the example below:
```

File: modules/ROOT/pages/guides/app-development/plugin.adoc [hunk 8]
```diff
@@ -0,0 +1,386 @@
...
+    required: ['enable', 'myKey1', 'myKey2'],
+    default: {
+        enable: true,
+        myKey1: 5000,
+        myKey2: ['127.0.0.1']
+    },
+}
+----
+
+This defines that the configuration options for the plugin can be provided in the following way:
```
Comment (chris529) [inline modules/ROOT/pages/guides/app-development/plugin.adoc]:
```suggestion
This defines that the configuration options for the plugin can be provided in the following manner:
```

File: modules/ROOT/pages/guides/app-development/plugin.adoc [hunk 9]
```diff
@@ -0,0 +1,386 @@
...
+
+== Defining the plugin logic
+
+The `load()` function of a plugin contains the plugin logic that is executed, when the plugin is initialized.
+
+It can be used to retrieve, mutate, store and/or publish data in a specific way, depending on the purpose of the plugin.
+
+The `unload()` method contains the logic that needs to be executed to unload the plugin correctly.
+
+The `channel`, which is available inside of the `load()` function, allows to access the xref:{url_rpc_endpoints}[] in order to subscribe to events or to invoke certain actions within the application, to retrieve the desired data.
```
Comment (chris529) [inline modules/ROOT/pages/guides/app-development/plugin.adoc]:
```suggestion
The `channel`, which is available inside of the `load()` function, allows access to the xref:{url_rpc_endpoints}[] in order to subscribe to events or to invoke certain actions within the application, to retrieve the desired data.
```

File: modules/ROOT/pages/guides/app-development/plugin.adoc [hunk 10]
```diff
@@ -0,0 +1,386 @@
...
+Similar to modules, plugins expose `actions` and `events`, which are interfaces that allow other plugins or external services to interact with the plugin.
+
+In this example, two events are added:
+
+* `newDelegate`, which is published in the `load()` function, when a new delegate is registered in ther network.
+* `timestamp`, which is published in the `load()` function, when a new block is added to the blockchain.
+
+In addition, two actions are added:
+
+* If `getKnownTimestamp` is invoked, it returns the list of timestamps of the blocks that where added to the chain, while the plugin was active.
```
Comment (chris529) [inline modules/ROOT/pages/guides/app-development/plugin.adoc]:
```suggestion
* If `getKnownTimestamp` is invoked, it returns the list of timestamps of the blocks that were added to the chain, while the plugin was active.
```

File: modules/ROOT/pages/guides/app-development/plugin.adoc [hunk 11]
```diff
@@ -0,0 +1,386 @@
...
+    this._knownTimestamps = [];
+  }
+}
+
+module.exports = MyPlugin;
+----
+
+== Registering the plugin with the application
+
+The last thing needed to do is to register the newly created module in the application:
```
Comment (chris529) [inline modules/ROOT/pages/guides/app-development/plugin.adoc]:
```suggestion
Finally, it is required to register the newly created module in the application:
```

**Review by chris529 — COMMENTED** [review]
Looks good, just a few very minor points 👍

**Review by chris529 — APPROVED** [review]
LGTM!

**Pull request merged**